module/mempool/consensus/exec_fork_suppressor.go

```diff
@@ -0,0 +1,345 @@
+	result := irSeal.IncorporatedResult.Result
+
+	initialState, ok := result.InitialStateCommit()
+	if !ok || len(initialState) < 1 {
+		scjson, err := json.Marshal(irSeal)
+		if err != nil {
+			return err
+		}
+		s.log.Error().
+			Str("seal", string(scjson)).
+			Msg("seal's execution result has no InitialStateCommit")
+		return engine.NewInvalidInputErrorf("seal's execution result has no InitialStateCommit: %x", result.ID())
+	}
+
+	finalState, ok := result.FinalStateCommitment()
+	if !ok || len(finalState) < 1 {
+		scjson, err := json.Marshal(irSeal)
+		if err != nil {
+			return err
+		}
+		s.log.Error().
+			Str("seal", string(scjson)).
+			Msg("seal's execution result has no FinalStateCommit")
+		return engine.NewInvalidInputErrorf("seal's execution result has no FinalStateCommit: %x", result.ID())
+	}
+
+	return nil
+}
+
+// getArbitraryElement picks and returns an arbitrary element from the set. Returns nil, if set is empty.
+func getArbitraryElement(set sealSet) *flow.IncorporatedResultSeal {
+	for _, seal := range set {
+		return seal
+	}
+	return nil
+}
+
+// enforceConsistentStateTransitions checks whether the execution results in the seals
+// have matching state transitions. If a fork in the execution state is detected:
+//   * wrapped mempool is cleared
+//   * internal execForkDetected flag is ste to true
+//   * the new value of execForkDetected is persisted to data base
+// and ExecutionForkErr (sentinel error) is returned
+func (s *ExecForkSuppressor) enforceConsistentStateTransitions(irSeal, irSeal2 *flow.IncorporatedResultSeal) error {
+	if irSeal.IncorporatedResult.Result.ID() == irSeal2.IncorporatedResult.Result.ID() {
+		// happy case: candidate seals are for the same result
+		return nil
+	}
+	// the results for the seals have different IDs (!)
+	// => check whether initial and final state match in both seals
+
+	sc1json, err := json.Marshal(irSeal)
+	if err != nil {
+		return fmt.Errorf("failed to marshal candidate seal to json: %w", err)
+	}
+	sc2json, err := json.Marshal(irSeal2)
+	if err != nil {
+		return fmt.Errorf("failed to marshal candidate seal to json: %w", err)
+	}
+	log := s.log.With().
+		Str("seal_1", string(sc1json)).
+		Str("seal_2", string(sc2json)).
+		Logger()
+
+	// unsafe: we assume validity of states has been checked before
+	irSeal1InitialState, _ := irSeal.IncorporatedResult.Result.InitialStateCommit()
+	irSeal1FinalState, _ := irSeal.IncorporatedResult.Result.FinalStateCommitment()
+	irSeal2InitialState, _ := irSeal2.IncorporatedResult.Result.InitialStateCommit()
+	irSeal2FinalState, _ := irSeal2.IncorporatedResult.Result.FinalStateCommitment()
+
+	if !bytes.Equal(irSeal1InitialState, irSeal2InitialState) || !bytes.Equal(irSeal1FinalState, irSeal2FinalState) {
+		log.Error().Msg("inconsistent seals for the same block")
+		s.seals.Clear()
+		s.execForkDetected = true
+		err := operation.RetryOnConflict(s.db.Update, operation.UpdateExecutionForkDetected(true))
```

Comment:
> storing the "evidence"
👍 implemented

> if the key exists [...] crash the node.
👍 implemented. For context:
* what we do in case we detect conflicting seals is now encapsulated in https://github.com/onflow/flow-go/blob/33de9d9fc417ea0319aabe3e35ed6e461d9d070d/module/mempool/consensus/exec_fork_actor.go#L12-L14 This allows us to easily replace the logic (and potentially keep going, just without sealing)